pkg/assembler/backends/testing/backend.go

```diff
@@ -45,11 +68,97 @@ func (c *demoClient) Packages(ctx context.Context, pkgSpec *model.PkgSpec) ([]*m
 	return packages, nil
 }
 
-func filterNamespace(pkg *model.Package, pkgSpec *model.PkgSpec) *model.Package {
+func (c *demoClient) Sources(ctx context.Context, sourceSpec *model.SourceSpec) ([]*model.Source, error) {
+	var sources []*model.Source
+	for _, s := range c.sources {
+		if sourceSpec.Type == nil || s.Type == *sourceSpec.Type {
+			newSource, err := filterSourceNamespace(s, sourceSpec)
+			if err != nil {
+				return nil, err
+			}
+			if newSource != nil {
+				sources = append(sources, newSource)
+			}
+		}
+	}
+	return sources, nil
+}
+
+func (c *demoClient) Cve(ctx context.Context, cveSpec *model.CVESpec) ([]*model.Cve, error) {
+	var cve []*model.Cve
+	for _, s := range c.cve {
+		if cveSpec.Year == nil || s.Year == *cveSpec.Year {
+			newCve, err := filterCVEID(s, cveSpec)
+			if err != nil {
+				return nil, err
+			}
+			if newCve != nil {
+				cve = append(cve, newCve)
+			}
+		}
+	}
+	return cve, nil
+}
+
+func (c *demoClient) Ghsa(ctx context.Context, ghsaSpec *model.GHSASpec) ([]*model.Ghsa, error) {
+	var ghsa []*model.Ghsa
+	for _, g := range c.ghsa {
+		newGHSA, err := filterGHSAID(g, ghsaSpec)
+		if err != nil {
+			return nil, err
+		}
+		if newGHSA != nil {
+			ghsa = append(ghsa, newGHSA)
+		}
+	}
+	return ghsa, nil
+}
+
+func (c *demoClient) Osv(ctx context.Context, osvSpec *model.OSVSpec) ([]*model.Osv, error) {
+	var osv []*model.Osv
+	for _, o := range c.osv {
+		newOSV, err := filterOSVID(o, osvSpec)
+		if err != nil {
+			return nil, err
+		}
+		if newOSV != nil {
+			osv = append(osv, newOSV)
+		}
+	}
+	return osv, nil
+}
+
+func (c *demoClient) Artifacts(ctx context.Context, artifactSpec *model.ArtifactSpec) ([]*model.Artifact, error) {
+	var artifacts []*model.Artifact
+	for _, a := range c.artifacts {
+		if artifactSpec.Digest == nil && artifactSpec.Algorithm == nil {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest != nil && artifactSpec.Algorithm == nil && a.Digest == *artifactSpec.Digest {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest == nil && artifactSpec.Algorithm != nil && a.Algorithm == *artifactSpec.Algorithm {
+			artifacts = append(artifacts, a)
+		} else if artifactSpec.Digest != nil && artifactSpec.Algorithm != nil && a.Algorithm == *artifactSpec.Algorithm && a.Digest == *artifactSpec.Digest {
+			artifacts = append(artifacts, a)
+		}
```

Comment:
i think we need to decide on the behavior here on cap sensitivity both for the protocol or the digest. We could follow what the intoto recommendation is and do lower case everything and have that be checked on ingestion time, we would want to specify it here and probably have the testing backend bootstrapped artifacts follow that convention.

Reply:
Yup, I agree to make it all lowercase. I will enforce lowercase in the testing.

Reply:
+1, we need to lowercase both when receiving the query and on ingesting.

Reply:
addressed PTAL!